### PR TITLE
bpo-36227: ElementTree.tostring() default_namespace argument

### PR DESCRIPTION
**Pull request by codeape2**
https://bugs.python.org/issue36227

### CODE REVIEW

**Comment by codeape2** [comment]
This PR also fixes bpo-31256 (https://bugs.python.org/issue31256): xml.etree.ElementTree: add support for doctype in tostring method

File: Doc/library/xml.etree.elementtree.rst [hunk 1]
```diff
@@ -594,19 +594,23 @@ Functions
 
 
 .. function:: tostring(element, encoding="us-ascii", method="xml", *, \
-                       short_empty_elements=True)
+                       default_namespace=None, short_empty_elements=True)
```
Comment (serhiy-storchaka) [inline Doc/library/xml.etree.elementtree.rst]: Add xml_declaration.

File: Doc/library/xml.etree.elementtree.rst [hunk 2]
```diff
@@ -594,19 +594,23 @@ Functions
...
 
    .. versionadded:: 3.4
       The *short_empty_elements* parameter.
 
+   .. versionadded:: 3.8
+      The *xml_declaration* and *default_namespace* parameters.
+
 
 .. function:: tostringlist(element, encoding="us-ascii", method="xml", *, \
                            short_empty_elements=True)
```
Comment (serhiy-storchaka) [inline Doc/library/xml.etree.elementtree.rst:618]: Add xml_declaration and default_namespace.

File: Lib/xml/etree/ElementTree.py [hunk 3]
```diff
@@ -1124,13 +1125,15 @@ def tostring(element, encoding=None, method=None, *,
...
-    be one of "xml" (default), "html", "text" or "c14n".
+    be one of "xml" (default), "html", "text" or "c14n", *default_namespace*
+    sets the default XML namespace (for "xmlns").
 
     Returns an (optionally) encoded string containing the XML data.
 
     """
     stream = io.StringIO() if encoding == 'unicode' else io.BytesIO()
-    ElementTree(element).write(stream, encoding, method=method,
+    ElementTree(element).write(stream, encoding, xml_declaration=xml_declaration,
```
Comment (serhiy-storchaka) [inline Lib/xml/etree/ElementTree.py]: Too long lines.

File: Lib/xml/etree/ElementTree.py [hunk 4]
```diff
@@ -1152,10 +1155,12 @@ def tell(self):
         return len(self.lst)
 
 def tostringlist(element, encoding=None, method=None, *,
+                 xml_declaration=None, default_namespace=None,
                  short_empty_elements=True):
     lst = []
     stream = _ListDataStream(lst)
-    ElementTree(element).write(stream, encoding, method=method,
+    ElementTree(element).write(stream, encoding, xml_declaration=xml_declaration,
```
Comment (serhiy-storchaka) [inline Lib/xml/etree/ElementTree.py]: Too long lines.

File: Misc/NEWS.d/next/Library/2019-03-07-20-02-18.bpo-36227.i2Z1XR.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Added new keyword arguments `default_namespace` and `xml_declaration` to functions
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2019-03-07-20-02-18.bpo-36227.i2Z1XR.rst]:
```suggestion
Added support for keyword arguments `default_namespace` and `xml_declaration` in functions
```

File: Lib/test/test_xml_etree.py [hunk 6]
```diff
@@ -756,6 +756,53 @@ def test_writestring(self):
...
         self.assertEqual(ET.tostring(elem), b'<html><body>text</body></html>')
 
+    def test_tostring_default_namespace(self):
+        elem = ET.XML('<body xmlns="http://effbot.org/ns"><tag/></body>')
+        self.assertEqual(
+            ET.tostring(elem, encoding='unicode'),
+            '<ns0:body xmlns:ns0="http://effbot.org/ns"><ns0:tag /></ns0:body>'
+        )
+        self.assertEqual(
+            ET.tostring(elem, encoding='unicode', default_namespace='http://effbot.org/ns'),
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py:767]: This is the same namespace as in the original XML. What if use different default namespase? What if use an original XML without xmlns?

File: Lib/test/test_xml_etree.py [hunk 7]
```diff
@@ -756,6 +756,53 @@ def test_writestring(self):
...
+        )
+
+    def test_tostring_xml_declaration(self):
+        elem = ET.XML('<body><tag/></body>')
+        self.assertEqual(
+            ET.tostring(elem, encoding='unicode'),
+            '<body><tag /></body>'
+        )
+        self.assertEqual(
+            str(ET.tostring(elem, encoding='utf8', xml_declaration=True), encoding='utf8'),
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py]: Test the following cases:
```python
ET.tostring(elem)
ET.tostring(elem, encoding='UTF-8')
ET.tostring(elem, encoding='US-ASCII')
ET.tostring(elem, encoding='ISO-8859-81')
ET.tostring(elem, encoding='unicode')
ET.tostring(elem, xml_declaration=False)
ET.tostring(elem, encoding='UTF-8', xml_declaration=False)
ET.tostring(elem, encoding='US-ASCII', xml_declaration=False)
ET.tostring(elem, encoding='ISO-8859-81', xml_declaration=False)
ET.tostring(elem, encoding='unicode', xml_declaration=False)
ET.tostring(elem, xml_declaration=True)
ET.tostring(elem, encoding='UTF-8', xml_declaration=True)
ET.tostring(elem, encoding='US-ASCII', xml_declaration=True)
ET.tostring(elem, encoding='ISO-8859-81', xml_declaration=True)
ET.tostring(elem, encoding='unicode', xml_declaration=True)
```

File: Lib/test/test_xml_etree.py [hunk 8]
```diff
@@ -756,6 +756,53 @@ def test_writestring(self):
...
+            "<?xml version='1.0' encoding='utf8'?>\n<body><tag /></body>"
+        )
+
+    def test_tostringlist_default_namespace(self):
+        elem = ET.XML('<body xmlns="http://effbot.org/ns"><tag/></body>')
+        self.assertEqual(
+            ET.tostringlist(elem, encoding='unicode')[0:2],
+            ['<ns0:body', ' xmlns:ns0="http://effbot.org/ns"']
```
Comment (serhiy-storchaka) [inline Lib/test/test_xml_etree.py]: The exact splitting on chunks is an implementation detail. Test the value of `b''.join(ET.tostringlist(elem, ...))` instead.

**Review by serhiy-storchaka — COMMENTED** [review]
Also, it would be nice to add similar tests for the `write()` method.

**Comment by codeape2** [comment]
@serhiy-storchaka I will look into your comments after the weekend.

File: Lib/test/test_xml_etree.py (hunk 6 above)
Comment (codeape2) [reply Lib/test/test_xml_etree.py:767]: The case of "original XML without xmlns" raises an exception:

```
>>> elem = ET.XML('<body><tag/></body>')
>>> ET.tostring(elem, encoding='unicode', default_namespace='foobar')
Traceback (most recent call last):
...
ValueError: cannot use non-qualified names with default_namespace option
```

@serhiy-storchaka is this the expected behavior?

Comment (serhiy-storchaka) [reply Lib/test/test_xml_etree.py:767]: Since it raises an exception with the special error message, this is the expected behavior.

**Comment by codeape2** [comment]
@serhiy-storchaka I pushed commits addressing your comments.

I did not add any similar tests for .write() however, since .write() is in effect tested by tostring().

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
LGTM. @scoder, could you please take a look?

**Pull request merged**